fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/tree/JoinEquivalentPredicatePushDown.java

```diff
@@ -0,0 +1,136 @@
+// This file is licensed under the Elastic License 2.0. Copyright 2021-present, StarRocks Inc.
+package com.starrocks.sql.optimizer.rule.tree;
+
+import com.starrocks.sql.optimizer.OptExpression;
+import com.starrocks.sql.optimizer.OptExpressionVisitor;
+import com.starrocks.sql.optimizer.Utils;
+import com.starrocks.sql.optimizer.base.ColumnRefFactory;
+import com.starrocks.sql.optimizer.operator.logical.LogicalJoinOperator;
+import com.starrocks.sql.optimizer.operator.scalar.ColumnRefOperator;
+import com.starrocks.sql.optimizer.operator.scalar.ScalarOperator;
+import com.starrocks.sql.optimizer.rewrite.ScalarEquivalenceExtractor;
+import com.starrocks.sql.optimizer.rule.transformation.PushDownJoinPredicateBase;
+import com.starrocks.sql.optimizer.task.TaskContext;
+
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Set;
+
+public class JoinEquivalentPredicatePushDown implements TreeRewriteRule {
```

Comment:
add comment

Reply:
Can you explain why we use tree rewrite strategy. What about tranformation strategy. We derive predicate from join on condition and filter condition of one child in the join. Then we add a filter on the top this join. The further process are relied on our exist rules to pushdown the derived predicate to the other child of this join.